test/core/pipeline/features/test_run.py

```diff
@@ -5082,6 +5090,280 @@ def pipeline_that_converts_files(pipeline_class):
+    txt_converter = TextFileToDocument()
+    csv_converter = CSVToDocument()
+    json_converter = JSONConverter(content_key="content")
+    html_converter = HTMLToDocument()
+
+    DocumentJoiner_1 = DocumentJoiner()
+    joiner = DocumentJoiner()
+    DocumentJoiner_2 = DocumentJoiner()
+
+    pp = pipeline_class(max_runs_per_component=2)
+
+    pp.add_component("router", router)
+    pp.add_component("splitter", splitter)
+    pp.add_component("txt_converter", txt_converter)
+    pp.add_component("csv_converter", csv_converter)
+    pp.add_component("json_converter", json_converter)
+    pp.add_component("html_converter", html_converter)
+    pp.add_component("joiner", joiner)
+    pp.add_component("DocumentJoiner_1", DocumentJoiner_1)
+    pp.add_component("DocumentJoiner_2", DocumentJoiner_2)
+    pp.add_component("page_splitter", page_splitter)
+
+    pp.connect("router.text/plain", "txt_converter.sources")
+    pp.connect("router.application/json", "json_converter.sources")
+    pp.connect("router.text/csv", "csv_converter.sources")
+    pp.connect("router.text/html", "html_converter.sources")
+    pp.connect("txt_converter.documents", "joiner.documents")
+    pp.connect("json_converter.documents", "joiner.documents")
+    pp.connect("csv_converter.documents", "DocumentJoiner_1.documents")
+    pp.connect("html_converter.documents", "DocumentJoiner_1.documents")
+    pp.connect("joiner.documents", "splitter.documents")
+    pp.connect("DocumentJoiner_1.documents", "page_splitter.documents")
+    pp.connect("splitter.documents", "DocumentJoiner_2.documents")
+    pp.connect("page_splitter.documents", "DocumentJoiner_2.documents")
+
+    expected_html_doc = Document(content="Some content", meta={"file_type": "html"})
+    expected_txt_doc = Document(content=txt_data, meta={"file_type": "txt"})
+    expected_txt_split_doc = Document(
+        content=txt_data,
+        meta={
+            "file_type": "txt",
+            "source_id": expected_txt_doc.id,
+            "page_number": 1,
+            "split_id": 0,
+            "split_idx_start": 0,
+        },
+    )
+    expected_html_split_doc = Document(
+        content=expected_html_doc.content,
+        meta={
+            "file_type": "html",
+            "source_id": expected_html_doc.id,
+            "page_number": 1,
+            "split_id": 0,
+            "split_idx_start": 0,
+        },
+    )
+
+    return (
+        pp,
+        [
+            PipelineRunData(
+                inputs={"router": {"sources": sources}},
+                # HTML converter takes longer than TXT Converter and this is why the order of documents is not stable
+                # for AsyncPipeline. We test for ANY here to make the test pass.
+                # In real usage, if the user cares about the order of documents arriving at a lazy variadic component,
+                # the user should pick a different component (OutputAdapter) to combine the lists.
```

Comment:
Interesting point. Not relevant to the PR but for that to work you'd need a way to provide default values for the Jinja2 variables in the situation where not both a HTML and TXT document are created in a batch. I don't believe this is currently possible with the OutputAdapter.